src/FSharp.Core/seq.fs

```diff
@@ -1938,3 +1938,186 @@ module Seq =
             if i < index then
                 invalidArg "index" "index must be within bounds of the array"
         }
+
+    [<CompiledName("RandomShuffleWith")>]
+    let randomShuffleWith (random: Random) (source: seq<'T>) : seq<'T> =
+        checkNonNull "random" random
+        checkNonNull "source" source
+
+        let tempArray = toArray source
+        Microsoft.FSharp.Primitives.Basics.Random.shuffleArrayInPlaceWith random tempArray
+        ofArray tempArray
+
+    [<CompiledName("RandomShuffleBy")>]
+    let randomShuffleBy (randomizer: unit -> float) (source: seq<'T>) : seq<'T> =
+        checkNonNull "source" source
+
+        let tempArray = toArray source
+        Microsoft.FSharp.Primitives.Basics.Random.shuffleArrayInPlaceBy randomizer tempArray
+        ofArray tempArray
+
+    [<CompiledName("RandomShuffle")>]
+    let randomShuffle (source: seq<'T>) : seq<'T> =
+        randomShuffleWith ThreadSafeRandom.Shared source
+
+    [<CompiledName("RandomChoiceWith")>]
+    let randomChoiceWith (random: Random) (source: seq<'T>) : 'T =
+        checkNonNull "random" random
+        checkNonNull "source" source
+
+        let inputLength = source |> length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+
+        let i = random.Next(0, inputLength)
+        source |> item i
```

Comment:
`source` should probably be cached in this case as well, since it would otherwise be enumerated twice.

Reply:
Same explanation as above